packages/app-defaults/src/defaults/apis.ts

```diff
@@ -64,10 +64,19 @@ export const apis = [
   createApiFactory({
     api: discoveryApiRef,
     deps: { configApi: configApiRef },
-    factory: ({ configApi }) =>
-      UrlPatternDiscovery.compile(
-        `${configApi.getString('backend.baseUrl')}/api/{{ pluginId }}`,
-      ),
+    factory: ({ configApi }) => {
+      let baseUrl;
+      try {
+        // Try parsing the url relative to the current document origin, if it fails the URL is misformed.
+        baseUrl = new URL(
+          configApi.getString('backend.baseUrl'),
+          document.location.origin,
+        ).href.replace(/\/$/, '');
+      } catch (err) {
+        baseUrl = configApi.getString('backend.baseUrl');
```

Comment:
Hate to say it but I feel we may need to figure out something more magic to make this work for the general case. This is not the only place where `backend.baseUrl` is read in the frontend so I'm feeling that would cause quite a lot of breakage. I'm thinking one option might actually be to detect and override within the `AppManager` in `core-app-api`. Basically if the `backend.baseUrl` is a path then we'd prepend a config override that sets it to `origin + <path>`. I don't love that idea though, it's a quite specific override.

Other options include a broader effort to make sure we handle all existing usages of `backend.baseUrl`, or that we add/update some API that surfaces this information, rather than having consumers read from config directly. What do you think?

Reply:
Hmm, I think the DiscoveryApi + IdentityAuthInjectorFetchMiddleware are the only locations that actually should reference `backend.baseUrl` (at the moment).

Taking a look through the repository, only the `catalog-graphql` plugin has a ```await fetch(`${this.baseUrl}/api/catalog/entities`)``` line which should be a DiscoveryApi call. I took a quick look at https://github.com/RoadieHQ/roadie-backstage-plugins and saw a similar thing for a few of their plugins. I'd imagine there are a bunch of company-specific plugins that have a similar issue. 

I'd say patching the `backend.baseUrl` in `AppManager` should be enough. The number of places `backend.baseUrl` is used should really be kept to a minimum and we have the DiscoveryApi to abstract it away.